### DISCUSSION TITLE
How to start number of workers filtering by cucumber tags

### DISCUSSION THREAD

**Discussion by darnold24** [webdriverio/webdriverio#8237] [Q&A]
Hi there, how are you guys doing. Been using your framework for a number of years.

So the issue I am having is what I thought would have been addressed in this article, unless I am not understanding the difference between workers and sessions: https://webdriver.io/blog/2019/11/01/spec-filtering/. It states that sessions will automatically be filtered by cucumber tags.

In our wdio config, we have 

```
maxInstances: 10,
  runner: 'local',
  reporters: [
    [
      'spec',
      {
        realtimeReporting: true,
        addConsoleLogs: true,
      },
    ],
    [
      'cucumberjs-json',
      {
        jsonFolder: `\${__dirname}/cucumber/e2eResults`\,
        language: 'en',
      },
    ],
  ],
  services: [
    [
      'chromedriver',
      {
        logFileName: 'wdio-chromedriver.log', // default
        outputDir: 'driver-logs', // overwrites the config.outputDir
        args: ['--silent'],
      },
    ],
  ],
  capabilities: [
    {
      browserName: 'chrome',
      acceptInsecureCerts: true,
      'goog:chromeOptions': {
        args: [
          '--disable-gpu',
          '--window-size=1920,1080',
          '--user-agent="Chrome/80.0.3987.122"',
          '--no-sandbox',
          '--disable-dev-shm-usage',
          '--headless',
        ],
      },
    },
  ],
  specs: [[`${__dirname}/../../consumer/charter-net-pages/spectrum/e2e/cucumber/features/**/*.feature`]],
  cucumberOpts: {
    require: [
      `${__dirname}/../../consumer/charter-net-pages/spectrum/e2e/cucumber/stepDefinitions/**/*.steps.ts`,
      `${__dirname}/cucumber/stepDefinitions/**/*.steps.ts`,
      `${__dirname}/cucumber/utils/hooks.ts`,
    ], // <string[]> (file/dir) require files before executing features
    backtrace: false, // <boolean> show full backtrace for errors
    requireModule: [], // <string[]> ("module") require MODULE files (repeatable)
    dryRun: false, // <boolean> invoke formatters without executing steps
    failFast: false, // <boolean> abort the run on first failure
    snippets: true, // <boolean> hide step definition snippets for pending steps
    source: true, // <boolean> hide source uris
    strict: false, // <boolean> fail if there are any undefined or pending steps
    tagExpression: '@runThisOnly', // <string> (expression) only execute the features or scenarios with tags matching the expression
    // tags: ['@sanity'],
    timeout: 60000, // <number> timeout for step definitions
    ignoreUndefinedDefinitions: false, // <boolean> Enable this config to treat undefined definitions as warnings.
  },
```
There are only two tests with @runOnlyThis, but it is creating 276 workers according to the console log. That would be the number of features we have in total, it is not filtering the number of workers by the cucumber tags.
```
Running Tests now...

Execution of 276 workers started at 2022-04-18T19:12:53.797Z
```

After about 25 minutes or it will eventually start, I think after it tries to start that many workers.

The only alternative we found to only run worker for the 2 tests was adding an extra outer array in the specs value:

```
[[`${__dirname}/../../consumer/charter-net-pages/spectrum/e2e/cucumber/features/**/*.feature`]]
```

but then we are getting a cucumber html report that doesn't separate the tests by feature file. @runThisOnly

<img width="590" alt="Screen Shot 2022-04-18 at 1 32 35 PM" src="https://user-images.githubusercontent.com/10222046/163865590-55792663-8c2b-4fa1-9164-c6ea64a3608d.png">

Anyway, is the framework working as expected, or is this a webdriver io issue?

**Comment by erwinheitzman** [comment]
Hi there, from what I have gathered it should indeed not spawn that many workers as this should have been tackled in this PR https://github.com/webdriverio/webdriverio/pull/4531 from what I can tell.

Seems like this could be a bug.

**Comment by darnold24** [comment]
Should I report as a bug?

**Reply by christian-bromann** [reply]
This could be a regression coming from the latest update to Cucumber v7. Please raise an issue with a minmal reproducible example, thanks!

**Reply by pfyod** [reply]
We stumbled on the same issue and reported the bug: #11994.

**Comment by SerhiiMaksymiv** [comment]
Any updates on this?
We are facing same issue, v8 version of wdio starts 400+ workers, which corresponds to number of feature files. We believe because of this our machine crashing with ‘js heap memory’.
Increasing heap size doesn’t help.

**Reply by christian-bromann** [reply]
@SerhiiMaksymivQAMadness if you can point out memory leaks in certain reporters or other place, please let us know. Any contributions are much appreciated.
